app/static/playlist.js

```diff
@@ -105,54 +142,107 @@ export default class PlaylistLabelRenderer {
+    }
+  }
+
+  jumpToLabel(label_id) {
+    console.log('jump to label', label_id);
+    // Update the current state
+    this.state.currentLabelId = label_id;
+    const items = this.state.items;
+
+    // make a list of items that starts with the new label
+    const upcoming_items = [];
+    const start_index = items.findIndex(function(item) {
+      return item.label && item.label.id === label_id;
+    });
+    for (let i=0; i<items.length; i++) {
+      upcoming_items.push(items[(start_index + i) % items.length]);
+    };
+
+    this.state.upcomingItems = upcoming_items;
+
+    // update label content
+    this.updateMainLabelContent(upcoming_items[0]);
+    if (items.length > 1) {
+      this.state.nextLabelId = upcoming_items[1].label.id;
+      this.updateUpNextContent(upcoming_items);
+    }
+  }
+
+  updateMainLabelContent(item) {
+    const label = item.label;
+    // Update the label fields with the currently playing data
+    const title = document.getElementById("title").innerHTML = label.title;
+    this.addTitleAnnotation(label.work);
+    document.getElementById("subtitles").innerHTML = label.subtitles;
+    document.getElementById("content0").innerHTML = label.columns[0].content;
+    document.getElementById("content1").innerHTML = label.columns[1].content;
+    document.getElementById("content2").innerHTML = label.columns[2].content;
+
+    if (label.work.is_context_indigenous) {
+      document.getElementById("indigenous").className = "indigenous indigenous_active";
+    } else {
+      document.getElementById("indigenous").className = "indigenous";
+    }
+  }
+
+  updateUpNextContent(items) {
+    // Update up next label
+    const item = items[1];
+    document.querySelector("#next_title").innerHTML = item.label.title;
+
+    for (let i=1; i<items.length; i++) {
+      const label = items[i].label;
+      const id = "#up_next_label_" + i;
+      try {
+        document.querySelector(id + " .title").innerHTML = i + '. ' + label.title;
+        document.querySelector(id + " .subtitles").innerHTML = label.subtitles;
+      } catch(err) {}
+    }
+  }
+
+  updateProgress() {
+    const progressBar = document.getElementById("progress-bar");
+    const playbackPosition = this.state.playbackPosition;
+    progressBar.style.width = `${playbackPosition * 100}%`;
+
+    const items = this.state.upcomingItems;
+
+    // calculate time to wait times for the upcoming videos;
+    let time_to_wait = items[0].video.duration_secs * (1.0 - playbackPosition);
+    for (let i=1; i<items.length; i++) {
+      const label = items[i].label;
+
+      const num_minutes = parseInt(Math.round(time_to_wait / 60.0));
+      let unit = ' minute';
+      if (num_minutes != 1) unit += 's';
+
+      const id = "#up_next_label_" + i;
+      try {
+        document.querySelector(id + " .time_to_wait").innerHTML = num_minutes + unit;
+      } catch(err) {}
+      time_to_wait += items[i].video.duration_secs;
+    }
+  }
+
+  autoUpdateProgress(slf) {
```

Comment:
I don't mind this for dev mode, but I think it should have to be explicitly turned on. My progress bar jumps backwards in time for some videos, so it's not great.

Reply:
I have turned it off, unless the 'ignore_media_player' flag is set. I haven't yet implemented a clean way to set it - it is always off in this repo, and always on in XOS preview.